File: scripts/generate-date-courses.ts
/**
 * 데이트 코스 생성 스크립트
 * Supabase DB의 places 데이터를 기반으로 거리 계산을 통해 데이트 코스를 생성합니다.
 */

import { createClient } from "@supabase/supabase-js"

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL || "https://dyomownljgsbwaxnljau.supabase.co"
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || ""

if (!SUPABASE_SERVICE_KEY) {
  console.error("SUPABASE_SERVICE_ROLE_KEY 환경 변수가 필요합니다.")
  process.exit(1)
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY)

// 지역 코드 매핑
const AREA_CODE_MAP: Record<number, string> = {
  1: "서울",
  2: "인천",
  3: "대전",
  4: "대구",
  5: "광주",
  6: "부산",
  7: "울산",
  8: "세종",
  31: "경기",
  32: "강원",
  33: "충북",
  34: "충남",
  35: "경북",
  36: "경남",
  37: "전북",
  38: "전남",
  39: "제주",
}

/**
 * 두 지점 간의 거리를 계산 (Haversine 공식)
 * @returns 거리 (km)
 */
function calculateDistance(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const R = 6371 // 지구 반경 (km)
  const dLat = ((lat2 - lat1) * Math.PI) / 180
  const dLng = ((lng2 - lng1) * Math.PI) / 180
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLng / 2) *
      Math.sin(dLng / 2)
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
  return R * c
}

/**
 * 데이트 코스에 적합한 장소 필터링
 */
function filterDatePlaces(places: any[]): any[] {
  return places.filter((place) => {
    // course_type에 'date'가 포함되어 있거나
    // CAFE, FOOD 타입이거나
    // MUSEUM 타입이면서 평점이 4.0 이상인 경우
    return (
      (place.course_type && place.course_type.includes("date")) ||
      place.type === "CAFE" ||
      place.type === "FOOD" ||
      (place.type === "MUSEUM" && parseFloat(place.rating || "0") >= 4.0)
    )
  })
}

/**
 * 거리 기반으로 최적의 코스 경로 생성 (Nearest Neighbor 알고리즘)
 */
function createOptimalRoute(places: any[], maxDistance: number = 5): any[] {
  if (places.length === 0) return []

  const route: any[] = []
  const visited = new Set<string>()
  let currentPlace = places[0] // 시작점

  route.push(currentPlace)
  visited.add(currentPlace.id)

  while (route.length < places.length) {
    let nearestPlace: any = null
    let nearestDistance = Infinity

    for (const place of places) {
      if (visited.has(place.id)) continue

      const distance = calculateDistance(
        parseFloat(currentPlace.lat),
        parseFloat(currentPlace.lng),
        parseFloat(place.lat),
        parseFloat(place.lng)
      )

      if (distance < nearestDistance && distance <= maxDistance) {
        nearestDistance = distance
        nearestPlace = place
      }
    }

    if (!nearestPlace) {
      // 더 이상 가까운 장소가 없으면 종료
      break
    }

    route.push(nearestPlace)
    visited.add(nearestPlace.id)
    currentPlace = nearestPlace
  }

  return route
}

/**
 * 지역별 데이트 코스 생성
 */
async function generateDateCourses() {
  console.log("데이트 코스 생성 시작...")

  // 지역별로 places 조회
  const { data: placesData, error: placesError } = await supabase
    .from("places")
    .select("id, name, lat, lng, type, rating, price_level, address, area_code, sigungu_code, course_type, image_url")
    .not("lat", "is", null)
    .not("lng", "is", null)
    .order("area_code")
    .order("sigungu_code")
    .order("rating", { ascending: false, nullsFirst: false })

  if (placesError) {
    console.error("Places 조회 오류:", placesError)
    return
  }

  if (!placesData || placesData.length === 0) {
    console.error("Places 데이터가 없습니다.")
    return
  }

  console.log(`총 ${placesData.length}개의 장소를 조회했습니다.`)

  // 데이트 코스에 적합한 장소 필터링
  const datePlaces = filterDatePlaces(placesData)
  console.log(`데이트 코스에 적합한 장소: ${datePlaces.length}개`)

  // 지역별로 그룹화
  const placesByRegion: Record<string, any[]> = {}

  for (const place of datePlaces) {
    const region = AREA_CODE_MAP[place.area_code] || "기타"
    if (!placesByRegion[region]) {
      placesByRegion[region] = []
    }
    placesByRegion[region].push(place)
  }

  console.log(`\n지역별 장소 수:`)
  for (const [region, places] of Object.entries(placesByRegion)) {
    console.log(`  ${region}: ${places.length}개`)
  }

  // 지역별로 코스 생성
  const courses: any[] = []

  for (const [region, places] of Object.entries(placesByRegion)) {
    if (places.length < 3) {
      console.log(`\n${region}: 장소가 너무 적어 코스를 생성하지 않습니다. (${places.length}개)`)
      continue
    }

    // 시군구별로 세분화하여 코스 생성
    const placesBySigungu: Record<number, any[]> = {}
    for (const place of places) {
      const sigungu = place.sigungu_code || 0
      if (!placesBySigungu[sigungu]) {
        placesBySigungu[sigungu] = []
      }
      placesBySigungu[sigungu].push(place)
    }

    for (const [sigungu, sigunguPlaces] of Object.entries(placesBySigungu)) {
      if (sigunguPlaces.length < 3) continue

      // 거리 기반으로 최적 경로 생성
      const route = createOptimalRoute(sigunguPlaces, 5) // 최대 5km 이내

      if (route.length >= 3) {
        // 코스 제목 생성
        const courseTitle = `${region} 데이트 코스 ${courses.filter((c) => c.region === region).length + 1}`

        courses.push({
          title: courseTitle,
          region,
          course_type: "date",
          description: `${region}의 다양한 데이트 장소를 포함한 코스입니다. 카페, 맛집, 문화시설 등이 포함되어 있습니다.`,
          image_url: route[0]?.image_url || null,
          place_count: route.length,
          area_code: places[0].area_code,
          sigungu_code: parseInt(sigungu),
          places: route,
        })

        console.log(`\n✓ ${courseTitle} 생성 완료 (${route.length}개 장소)`)
      }
    }
  }

  console.log(`\n총 ${courses.length}개의 데이트 코스를 생성했습니다.`)

  // DB에 저장
  console.log("\nDB에 저장 중...")

  for (const course of courses) {
    try {
      // 코스 생성
      const { data: courseData, error: courseError } = await supabase
        .from("courses")
        .insert({
          title: course.title,
          region: course.region,
          course_type: course.course_type,
          description: course.description,
          image_url: course.image_url,
          place_count: course.place_count,
          area_code: course.area_code,
          sigungu_code: course.sigungu_code,
        })
        .select()
        .single()

      if (courseError) {
        console.error(`코스 생성 오류 (${course.title}):`, courseError)
        continue
      }

      // 코스에 장소 연결
      const coursePlaces = course.places.map((place: any, index: number) => ({
        course_id: courseData.id,
        place_id: place.id,
        order_index: index,
      }))

      const { error: coursePlacesError } = await supabase.from("course_places").insert(coursePlaces)

      if (coursePlacesError) {
        console.error(`코스 장소 연결 오류 (${course.title}):`, coursePlacesError)
        // 코스는 생성되었지만 장소 연결 실패 시 코스 삭제
        await supabase.from("courses").delete().eq("id", courseData.id)
        continue
      }

      console.log(`✓ ${course.title} 저장 완료`)
    } catch (error) {
      console.error(`코스 저장 중 오류 (${course.title}):`, error)
    }
  }

  console.log("\n데이트 코스 생성 완료!")
}

// 스크립트 실행
generateDateCourses()
  .then(() => {
    console.log("작업 완료")
    process.exit(0)
  })
  .catch((error) => {
    console.error("오류 발생:", error)
    process.exit(1)
  })

File: scripts/generate_date_courses.py
#!/usr/bin/env python3
"""
데이트 코스 생성 스크립트
Supabase DB의 places 데이터를 기반으로 거리 계산을 통해 데이트 코스를 생성합니다.
"""

import os
import math
import random
from supabase import create_client, Client
from typing import List, Dict, Any

# 환경 변수에서 설정 가져오기
SUPABASE_URL = os.getenv("NEXT_PUBLIC_SUPABASE_URL", "https://dyomownljgsbwaxnljau.supabase.co")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY", "")

if not SUPABASE_KEY:
    print("오류: SUPABASE_SERVICE_ROLE_KEY 환경 변수가 필요합니다.")
    exit(1)

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

# 지역 코드 매핑
AREA_CODE_MAP = {
    1: "서울", 2: "인천", 3: "대전", 4: "대구", 5: "광주",
    6: "부산", 7: "울산", 8: "세종", 31: "경기", 32: "강원",
    33: "충북", 34: "충남", 35: "경북", 36: "경남",
    37: "전북", 38: "전남", 39: "제주"
}

def calculate_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """두 지점 간의 거리를 계산 (Haversine 공식) - km 단위"""
    R = 6371  # 지구 반경 (km)
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)
    a = (
        math.sin(d_lat / 2) ** 2 +
        math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
        math.sin(d_lng / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c

def filter_date_places(places: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """데이트 코스에 적합한 장소 필터링"""
    filtered = []
    for place in places:
        course_type = place.get("course_type", [])
        place_type = place.get("type", "")
        rating = float(place.get("rating", 0) or 0)
        
        # course_type에 'date'가 포함되어 있거나
        # CAFE, FOOD 타입이거나
        # MUSEUM 타입이면서 평점이 4.0 이상인 경우
        if (
            (course_type and "date" in course_type) or
            place_type == "CAFE" or
            place_type == "FOOD" or
            (place_type == "MUSEUM" and rating >= 4.0)
        ):
            filtered.append(place)
    return filtered

def create_optimal_route(places: List[Dict[str, Any]], max_distance: float = 5.0) -> List[Dict[str, Any]]:
    """거리 기반으로 최적의 코스 경로 생성 (Nearest Neighbor 알고리즘)"""
    if not places:
        return []
    
    route = []
    visited = set()
    
    # 시작점 선택 (평점 높은 순)
    current_place = max(places, key=lambda p: float(p.get("rating", 0) or 0))
    route.append(current_place)
    visited.add(current_place["id"])
    
    # 최대 10개까지 추가
    while len(route) < min(10, len(places)):
        nearest_place = None
        nearest_distance = float("inf")
        
        for place in places:
            if place["id"] in visited:
                continue
            
            distance = calculate_distance(
                float(current_place["lat"]),
                float(current_place["lng"]),
                float(place["lat"]),
                float(place["lng"])
            )
            
            if distance < nearest_distance and distance <= max_distance:
                nearest_distance = distance
                nearest_place = place
        
        if not nearest_place:
            break
        
        route.append(nearest_place)
        visited.add(nearest_place["id"])
        current_place = nearest_place
    
    return route

def generate_date_courses():
    """지역별 데이트 코스 생성"""
    print("데이트 코스 생성 시작...")
    
    # 모든 places 조회
    response = supabase.table("places").select(
        "id, name, lat, lng, type, rating, price_level, address, "
        "area_code, sigungu_code, course_type, image_url"
    ).not_.is_("lat", "null").not_.is_("lng", "null").order("area_code").order("sigungu_code").order("rating", desc=True).limit(10000).execute()
    
    if not response.data:
        print("오류: Places 데이터를 조회할 수 없습니다.")
        return
    
    places = response.data
    print(f"총 {len(places)}개의 장소를 조회했습니다.")
    
    # 데이트 코스에 적합한 장소 필터링
    date_places = filter_date_places(places)
    print(f"데이트 코스에 적합한 장소: {len(date_places)}개")
    
    # 지역별로 그룹화
    places_by_region: Dict[str, List[Dict[str, Any]]] = {}
    for place in date_places:
        area_code = place.get("area_code")
        if not area_code:
            continue
        region = AREA_CODE_MAP.get(area_code, "기타")
        if region not in places_by_region:
            places_by_region[region] = []
        places_by_region[region].append(place)
    
    print(f"\n지역별 장소 수:")
    for region, region_places in places_by_region.items():
        print(f"  {region}: {len(region_places)}개")
    
    # 지역별로 코스 생성
    courses_created = 0
    
    for region, region_places in places_by_region.items():
        if len(region_places) < 3:
            print(f"\n{region}: 장소가 너무 적어 코스를 생성하지 않습니다. ({len(region_places)}개)")
            continue
        
        # 시군구별로 세분화
        places_by_sigungu: Dict[int, List[Dict[str, Any]]] = {}
        for place in region_places:
            sigungu = place.get("sigungu_code") or 0
            if sigungu not in places_by_sigungu:
                places_by_sigungu[sigungu] = []
            places_by_sigungu[sigungu].append(place)
        
        course_count = 0
        for sigungu, sigungu_places in places_by_sigungu.items():
            if len(sigungu_places) < 3:
                continue
            
            # 거리 기반으로 최적 경로 생성
            route = create_optimal_route(sigungu_places, max_distance=5.0)
            
            if len(route) >= 3:
                course_count += 1
                course_title = f"{region} 데이트 코스 {course_count}"
                
                try:
                    # 코스 생성
                    course_data = {
                        "title": course_title,
                        "region": region,
                        "course_type": "date",
                        "description": f"{region}의 다양한 데이트 장소를 포함한 코스입니다. 카페, 맛집, 문화시설 등이 포함되어 있습니다.",
                        "image_url": route[0].get("image_url"),
                        "place_count": len(route),
                        "area_code": region_places[0].get("area_code"),
                        "sigungu_code": sigungu
                    }
                    
                    course_result = supabase.table("courses").insert(course_data).execute()
                    course_id = course_result.data[0]["id"]
                    
                    # 코스에 장소 연결
                    course_places_data = [
                        {
                            "course_id": course_id,
                            "place_id": place["id"],
                            "order_index": idx
                        }
                        for idx, place in enumerate(route)
                    ]
                    
                    supabase.table("course_places").insert(course_places_data).execute()
                    
                    print(f"✓ {course_title} 생성 완료 ({len(route)}개 장소)")
                    courses_created += 1
                except Exception as e:
                    print(f"✗ {course_title} 생성 실패: {e}")
        
        if course_count == 0:
            print(f"\n{region}: 코스를 생성할 수 있는 충분한 장소가 없습니다.")
    
    print(f"\n총 {courses_created}개의 데이트 코스를 생성했습니다.")
    print("데이트 코스 생성 완료!")

if __name__ == "__main__":
    generate_date_courses()

